File: frontend/src/types/bulk_evaluation.ts
/**
 * TypeScript Interfaces for Part 9: Bulk Verification & Batch Processing
 */

export interface BulkEvaluationSummaryCounts {
  total: number;
  processed: number;
  successful: number;
  failed: number;
  review_required: number;
  critical_findings: number;
  remaining: number;
  progress_percentage: number;
}

export interface BulkEvaluationJobCreateResponse {
  job_id: string;
  tender_id: string;
  status: string;
  total_bids: number;
  message: string;
  created_at: string;
}

export interface BulkEvaluationJobStatusResponse {
  id: string;
  organization_id: string;
  tender_id: string;
  tender_number?: string | null;
  tender_title?: string | null;
  status: "QUEUED" | "RUNNING" | "COMPLETED" | "PARTIALLY_COMPLETED" | "FAILED" | "CANCELLED" | string;
  counts: BulkEvaluationSummaryCounts;
  started_by_name?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
  error_summary?: Record<string, any> | null;
}

export interface BulkEvaluationJobItem {
  id: string;
  job_id: string;
  bid_id: string;
  bid_number?: string | null;
  bidder_name?: string | null;
  status: "QUEUED" | "RUNNING" | "SUCCESS" | "FAILED" | "REVIEW_REQUIRED" | "SKIPPED" | string;
  current_stage: "QUEUED" | "DOCUMENT_PROCESSING" | "VERIFICATION" | "COMPLIANCE" | "SCORING" | "RISK" | "COMPLETED" | "FAILED" | "SKIPPED" | string;
  document_processing_status: string;
  verification_status: string;
  compliance_status: string;
  score_status: string;
  risk_status: string;
  final_score?: number | null;
  risk_level?: string | null;
  review_required: boolean;
  critical_findings_count: number;
  error_code?: string | null;
  error_message?: string | null;
  is_retryable: boolean;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
}

export interface BulkEvaluationJobItemsListResponse {
  items: BulkEvaluationJobItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface BulkEvaluationRetryResponse {
  job_id: string;
  retried_count: number;
  status: string;
  message: string;
}

export interface BulkEvaluationCancelResponse {
  job_id: string;
  status: string;
  message: string;
}

File: backend/app/schemas/bulk_evaluation.py
"""
Pydantic Schemas for Part 9: Bulk Verification & Batch Processing
Provides structured request payloads, response models, progress telemetry,
and paginated item listings for tender-level batch evaluation runs.
"""

import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class BulkEvaluationSummaryCounts(BaseModel):
    total: int = Field(0, description="Total eligible submitted bids in batch")
    processed: int = Field(0, description="Number of bids processed so far")
    successful: int = Field(0, description="Number of bids processed with full success")
    failed: int = Field(0, description="Number of bids with technical processing errors")
    review_required: int = Field(0, description="Number of bids requiring human review")
    critical_findings: int = Field(0, description="Number of bids with critical non-compliance")
    remaining: int = Field(0, description="Number of bids awaiting processing")
    progress_percentage: float = Field(0.0, description="Completion percentage (0.0 - 100.0)")


class BulkEvaluationJobCreateResponse(BaseModel):
    job_id: uuid.UUID
    tender_id: uuid.UUID
    status: str
    total_bids: int
    message: str
    created_at: datetime


class BulkEvaluationJobStatusResponse(BaseModel):
    id: uuid.UUID
    organization_id: uuid.UUID
    tender_id: uuid.UUID
    tender_number: Optional[str] = None
    tender_title: Optional[str] = None
    status: str
    counts: BulkEvaluationSummaryCounts
    started_by_name: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    error_summary: Optional[Dict[str, Any]] = None


class BulkEvaluationJobItemResponse(BaseModel):
    id: uuid.UUID
    job_id: uuid.UUID
    bid_id: uuid.UUID
    bid_number: Optional[str] = None
    bidder_name: Optional[str] = None
    status: str
    current_stage: str
    document_processing_status: str
    verification_status: str
    compliance_status: str
    score_status: str
    risk_status: str
    final_score: Optional[float] = None
    risk_level: Optional[str] = None
    review_required: bool = False
    critical_findings_count: int = 0
    error_code: Optional[str] = None
    error_message: Optional[str] = None
    is_retryable: bool = False
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime


class BulkEvaluationJobItemsListResponse(BaseModel):
    items: List[BulkEvaluationJobItemResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class BulkEvaluationRetryResponse(BaseModel):
    job_id: uuid.UUID
    retried_count: int
    status: str
    message: str


class BulkEvaluationCancelResponse(BaseModel):
    job_id: uuid.UUID
    status: str
    message: str
